// ngx-gravatar/esm2020/lib/gravatar.module.mjs
import { NgModule } from '@angular/core';
import { GRAVATAR_CONFIG_TOKEN } from './gravatar-config.token';
import { NgxGravatarDirective } from './ngx-gravatar.directive';
import * as i0 from "@angular/core";
export class GravatarModule {
    static forRoot(gravatarConfig) {
        return {
            ngModule: GravatarModule,
            providers: [
                {
                    provide: GRAVATAR_CONFIG_TOKEN,
                    useValue: gravatarConfig,
                },
            ],
        };
    }
}
GravatarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: GravatarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
GravatarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: GravatarModule, declarations: [NgxGravatarDirective], exports: [NgxGravatarDirective] });
GravatarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: GravatarModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: GravatarModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgxGravatarDirective],
                    exports: [NgxGravatarDirective],
                }]
        }] });
